compiler/rustc_transmute/src/layout/mod.rs

```diff
@@ -30,33 +30,47 @@ impl fmt::Debug for Byte {
 }
 
 pub(crate) trait Def: Debug + Hash + Eq + PartialEq + Copy + Clone {}
-pub trait Ref: Debug + Hash + Eq + PartialEq + Copy + Clone {}
+pub trait Ref: Debug + Hash + Eq + PartialEq + Copy + Clone {
+    fn min_align(&self) -> usize {
+        1
+    }
+
+    fn is_mutable(&self) -> bool {
+        false
+    }
+}
 
 impl Def for ! {}
 impl Ref for ! {}
 
 #[cfg(feature = "rustc")]
-pub(crate) mod rustc {
+pub mod rustc {
     use rustc_middle::mir::Mutability;
-    use rustc_middle::ty;
-    use rustc_middle::ty::Region;
-    use rustc_middle::ty::Ty;
+    use rustc_middle::ty::{self, Ty};
 
     /// A reference in the layout.
     #[derive(Debug, Hash, Eq, PartialEq, PartialOrd, Ord, Clone, Copy)]
     pub struct Ref<'tcx> {
-        lifetime: Region<'tcx>,
-        ty: Ty<'tcx>,
-        mutability: Mutability,
+        pub lifetime: ty::Region<'tcx>,
+        pub ty: Ty<'tcx>,
+        pub mutability: Mutability,
     }
 
-    impl<'tcx> super::Ref for Ref<'tcx> {}
+    impl<'tcx> super::Ref for Ref<'tcx> {
+        fn min_align(&self) -> usize {
+            // FIXME(bryangarza): Need to have access to a ctx: TyCtxt<'tcx> here
+            // super::tree::rustc::layout_of(ctx, self.ty).unwrap().align()
```

Comment:
Yeah, the main problem I have here is that the code is split by a feature flag that enables rustc dependencies, and here I only have access to a trait object `QueryContext` that is supposed to abstract `TyCtxt` so it's not directly used... IDK what the best way to deal with this is, downcasting isn't really possible because we don't have a static lifetime for `TyCtxt`, the other way I could handle it is just moving more code to use the rustc feature flag but at that point, we should maybe just get rid of the flag and always depend on rustc. (IDK why, but I think the original reason we have it is to be able to build `rustc_transmute` as its own crate independent of the compiler.

Reply:
Well then QueryContext would be the thing that has a layout-of method exposed on it, it seems?

But yeah, I don't think all this extra abstraction is actually helping the design of this crate. I'd personally rather this be designed with rustc in mind, then later if we want this to be a separate crate (like Chalk) then it can be re-abstracted out with a clearer understanding of what the API boundaries need to be.

Reply:
How about we just add an `align` *field* to `Ref`, and fill in that field during the`Tree::from_ty` process (where we have access to `TyCtxt`).